src/AppInstallerCLICore/ExecutionContext.cpp

```diff
@@ -79,43 +127,136 @@ namespace AppInstaller::CLI::Execution
+
+                wcex.style = CS_NOCLOSE;
+                wcex.lpfnWndProc = WindowMessageProcedure;
+                wcex.cbClsExtra = 0;
+                wcex.cbWndExtra = 0;
+                wcex.hInstance = hInstance;
+                wcex.lpszClassName = windowClass;
+
+                if (!RegisterClassEx(&wcex))
+                {
+                    LOG_LAST_ERROR_MSG("Failed registering window class");
+                    return;
+                }
+
+                m_windowHandle = wil::unique_hwnd(CreateWindow(
+                    windowClass,
+                    L"WingetMessageOnlyWindow",
+                    WS_OVERLAPPEDWINDOW,
+                    0, /* x */
+                    0, /* y */
+                    0, /* nWidth */
+                    0, /* nHeigth */
+                    NULL, /* hWndParent */
+                    NULL, /* hMenu */
+                    hInstance,
+                    NULL)); /* lpParam */
+
+                if (m_windowHandle == nullptr)
+                {
+                    LOG_LAST_ERROR_MSG("Failed creating window");
+                    return;
+                }
+
+                ShowWindow(m_windowHandle.get(), SW_HIDE);
+
+                // Force message queue to be created.
+                MSG msg;
+                PeekMessage(&msg, NULL, WM_USER, WM_USER, PM_NOREMOVE);
+                m_messageQueueReady.SetEvent();
+
+                // Message loop
+                BOOL getMessageResult;
+                while ((getMessageResult = GetMessage(&msg, m_windowHandle.get(), 0, 0)) != 0)
+                {
+                    if (getMessageResult == -1)
+                    {
+                        LOG_LAST_ERROR();
+                    }
+                    else
+                    {
+                        DispatchMessage(&msg);
+                    }
+                }
+            }
+
+#ifndef AICLI_DISABLE_TEST_HOOKS
+            wil::unique_event m_appShutdownEvent;
+#endif
+
             std::mutex m_contextsLock;
             std::vector<Context*> m_contexts;
+            wil::unique_event m_messageQueueReady;
+            wil::unique_hwnd m_windowHandle;
+            std::thread m_windowThread;
         };
 
-        void SetCtrlHandlerContext(bool add, Context* context)
+        void SetSignalTerminationHandlerContext(bool add, Context* context)
         {
             THROW_HR_IF(E_POINTER, context == nullptr);
 
             if (add)
             {
-                CtrlHandler::Instance().AddContext(context);
+                SignalTerminationHandler::Instance().AddContext(context);
             }
             else
             {
-                CtrlHandler::Instance().RemoveContext(context);
+                SignalTerminationHandler::Instance().RemoveContext(context);
+            }
+        }
+
+        // This starts getting called by CreateWindow.
+        // If it's a member function in SignalTerminationHandler we will have problems.

```

Comment:
If it is a static function in `SignalTerminationHandler` it should work just fine, and not need to be so far from the usage nor pre-declared.